src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,258 @@
+"""Low level control logic for actuating the rudder and the sail."""
+
+from abc import ABC, abstractmethod
+from math import atan2, copysign, cos, sin
+from typing import Tuple
+
+import numpy as np
+
+from boat_simulator.common.constants import RUDDER_MAX_ANGLE_RANGE, SAIL_MAX_ANGLE_RANGE
+from boat_simulator.common.types import Scalar
+from boat_simulator.common.utils import bound_to_180
+
+
+class ActuatorController(ABC):
+    """Abstract class for rudder and sail actuation mechanisms.
+
+    Attributes:
+        `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+        `time_step` (Scalar): Time taken per iteration given in seconds.
+        `control_speed` (Scalar): Speed of control angle change in degrees / second.
+        'running_error' (Scalar): Error between current and target control angle.
+    """
+
+    def __init__(
+        self,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        control_speed: Scalar,
+        running_error: Scalar,
+        max_angle_range: Tuple[Scalar, Scalar],
+    ):
+        """Initializes the class attributes. Note that this class cannot be directly instantiated.
+
+        Args:
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in seconds.
+            `control_speed` (Scalar): Speed of control angle change in degrees / second.
+            'running_error' (Scalar): Error between current and target control angle.
+        """
+        self.current_control_ang = current_control_ang
+        self.time_step = time_step
+        self.control_speed = control_speed
+        self.running_error = running_error
+        self.max_angle_range = max_angle_range
+
+    def update_state(self) -> bool:
+        """Updates the controller position iteratively based on control_speed * time_step, until
+        the control target angle is reached.
+
+        Args:
+            `max_angle_range` (tuple): Max angle range individual to each different controller
+
+
+        Returns:
+            False if target control angle was not reached, else
+            True if target control angle has been reached (running error = 0)
+        """
+        error_tracker = False
+        if np.isclose(self.running_error, 0, 0.01):
+            return True
+        else:
+            change = copysign(1, self.running_error) * self.control_speed * self.time_step
+
+        if abs(self.running_error) > abs(change):
+            next_control = max(self.current_control_ang + change, self.max_angle_range[0])
+            self.running_error -= change
+        else:
+            next_control = max(
+                self.current_control_ang + self.running_error, self.max_angle_range[0]
+            )
+            self.running_error = 0
+            error_tracker = True
+
+        self.current_control_ang = min(next_control, self.max_angle_range[1])
```

Comment:
Basically, the first if statement checks if the magnitude of the running error is greater than the magnitude of change. This determines whether a full "change" iteration can be completed within the next time step, or if the difference is less than what is required for one iteration. For example, if the running_error was 10, and our control speed per time step was 2, we know that we can change by a full iteration of 2 degrees. In contrast, if the running _error were 1.5, we would know to not actuate by 2 degrees to avoid overshooting the target control angle. 

In the loop itself, next_control stores the larger value between the current control angle + change or the index 1 of the RUDDER_MAX_ANGLE constant, which in this case is -45 degrees, bounding it above -45 degrees. I'm currently calculating running_error by target - current, meaning if the error is negative, the target is larger than the current and the change is negative. If the error is positive, that means the target angle is less than the current control, meaning our change should be positive. If the running error is negative and its change is negative, subtracting the change from the running error will result in running_error converging towards zero. In the else loop, we run the same next_control check, except we add whatever is remaining in the running error, which will be less than the change. Our running_error should be set to 0, and the error tracker tracks that we reached our target control angle, returning True. The last check bounds the angle value to be less than 45 degrees.

Hopefully that made some sense haha.